File: web/src/notifications/mentions.ts
// @mention tokens are the inline sibling of the fls:doc / fls:task card
// tokens (see components/reftokens.ts): a compact, text-safe pseudo-URL the
// chat composer inserts when you pick a member, stored verbatim in the
// plain-text body. The renderer swaps it for a highlighted "@Name" at display
// time, and the SERVER parses the same token to address the mention
// notification — so a mention costs no APS roster call at either end.
//
// Unlike the card tokens, a mention renders as inline text (a highlighted
// span), not a bordered card, so it lives here rather than in RefCard.

export interface MentionRef {
  userId: string
  name: string
}

export const MENTION_PREFIX = 'fls:user?'

// The character class is char-for-char the one every other fls: token uses
// (reftokens.ts), so trailing punctuation never sticks to a token and both
// the client and the Go parser split identically.
const MENTION_RE = /fls:user\?[A-Za-z0-9*\-._%&=+]+/g

export function encodeMention(m: MentionRef): string {
  const sp = new URLSearchParams()
  sp.set('id', m.userId)
  sp.set('name', m.name)
  return MENTION_PREFIX + sp.toString()
}

export function parseMention(token: string): MentionRef | null {
  if (!token.startsWith(MENTION_PREFIX)) return null
  const sp = new URLSearchParams(token.slice(MENTION_PREFIX.length))
  const userId = sp.get('id') ?? ''
  if (!userId) return null
  return { userId, name: sp.get('name') || '' }
}

export type MentionPart = { text: string } | { mention: MentionRef }

// splitMentions breaks a run of text into plain runs and mention tokens.
// Malformed tokens (no id) stay text. Used on the text parts that survive
// splitRefTokens, so cards and mentions compose without either eating the
// other's tokens.
export function splitMentions(text: string): MentionPart[] {
  const parts: MentionPart[] = []
  let last = 0
  for (const m of text.matchAll(MENTION_RE)) {
    const mention = parseMention(m[0])
    if (!mention) continue // malformed: leave as text
    if (m.index! > last) parts.push({ text: text.slice(last, m.index) })
    parts.push({ mention })
    last = m.index! + m[0].length
  }
  if (last < text.length) parts.push({ text: text.slice(last) })
  return parts.length ? parts : [{ text }]
}

File: notifications/mentions.go
package notifications

import (
	"net/url"
	"regexp"
	"strings"
)

// MentionToken is the inline card-token scheme for an @mention, mirroring the
// repo's other fls: tokens (fls:doc / fls:task / fls:job / fls:batch). It is
// authored by the composer (which knows the roster) and stored verbatim in
// the plain-text body, so the SERVER can resolve mention targets by parsing
// the body alone — no APS roster fetch, which matters under the per-minute
// cost quota. The client unfurls it to a highlighted "@Name" at render time.
const MentionToken = "fls:user"

// mentionRe matches a well-formed fls:user token. The character class is the
// same one web/src/components/reftokens.ts uses for every fls: token, so both
// ends split identically; malformed tokens simply don't match and stay text.
var mentionRe = regexp.MustCompile(`fls:user\?[A-Za-z0-9*\-._%&=+]+`)

// Mention is a resolved @mention: the target's stable id and the display name
// captured in the token.
type Mention struct {
	UserID string
	Name   string
}

// ParseMentions extracts the unique @mention targets from a message body,
// preserving first-seen order and dropping tokens without an id. Duplicate
// mentions of the same user collapse to one (you get one notification even if
// named twice in a message).
func ParseMentions(body string) []Mention {
	matches := mentionRe.FindAllString(body, -1)
	if len(matches) == 0 {
		return nil
	}
	out := make([]Mention, 0, len(matches))
	seen := make(map[string]bool, len(matches))
	for _, tok := range matches {
		q := tok[len(MentionToken)+1:] // drop "fls:user?"
		vals, err := url.ParseQuery(q)
		if err != nil {
			continue
		}
		id := strings.TrimSpace(vals.Get("id"))
		if id == "" || seen[id] {
			continue
		}
		seen[id] = true
		out = append(out, Mention{UserID: id, Name: strings.TrimSpace(vals.Get("name"))})
	}
	return out
}
